---
title: 'JASPAR 2022 motif discovery pipeline and motif annotation'
author: "Jaime A Castro-Mondragon"
date: 'Last Update: `r Sys.Date()`'
geometry: margin=1.5cm
output: 
  pdf_document
bibliography: References.bib
# biblio-style: apalike
csl: NAR.csl
link-citations: yes
header-includes: 
  \usepackage{float} \floatplacement{figure}{H} 
  \newcommand{\beginsupplement}{\setcounter{table}{0}
  \renewcommand{\thetable}{S\arabic{table}} \setcounter{figure}{0}
  \renewcommand{\thefigure}{S\arabic{figure}}}
  \usepackage{fullpage}
  \usepackage{pdflscape}
  \usepackage{inputenc} 
---

# Summary 

[**JASPAR**](http://jaspar.genereg.net) [@Fornes2020] is an open-access database of curated, non-redundant 
transcription factor (TF)-binding profiles stored as position frequency matrices 
(PFMs) for TFs across multiple species in six taxonomic groups: Vertebrates, Plants,
Fungi, Insects and Urochordata.


For JASPAR 2022 release we curated motifs derived from the following sources:

  - Human TF dimers motifs derived from SELEX [@Jolma2015]
  - Human TF motifs detected by SELEX in free DNA [@Zhu2018]
  - Human TF motifs detected by SELEX in nucleosomal DNA [@Zhu2018]
  - Human Zinc finger TF motifs derived from ChIP-seq [@Dogan2020]
  - Human Zinc finger TF motifs derived from ChIP-exo [@Dogan2020]
  - *Ciona robusta* motifs derived from SELEX [@Nitta2019]
  <!-- - Yeast motifs derived from ChIP-exo [] -->
  <!-- - *Cnidaria* motifs derived from ATAC-seq [@Hettinger2018] -->


# Requisites

- This motif discovery pipeline requires the regulatory sequences analysis tools [**RSAT**](http://rsat-tagc.univ-mrs.fr/rsat/) [@Nguyen2018] installed in your machine/server.

- [**R**](https://www.r-project.org/) >= 3.6 

___


## Pre-processing of human zinc finger motifs from Dogan 2020

The [**ChIP-seq**](https://www.biorxiv.org/content/biorxiv/early/2020/06/16/630756/DC9/embed/media-9.zip?download=true) and [**ChIP-exo**](https://www.biorxiv.org/content/biorxiv/early/2020/06/16/630756/DC10/embed/media-10.zip?download=true) motif libraries were downloaded from the 
[**supplementary material**](https://www.biorxiv.org/content/10.1101/630756v4.supplementary-material) from [@Dogan2020].

Each motif comes in a separated file in *cis-bp* motif format. The file name corresponds
to their uniprot ID.

Run the following command to map the correct TF name from the uniprot ID and convert
the motifs from *cis-bp* to *transfac* format, the latter allows to add annotations
and comments to the motifs.

Download and unzip the each zip file in a separated folder.


Command: 

- *motif_folder* : directory containing the *cis-bp* motifs
- *logo_folder*  : directory where the logos will be stored
- *motif_source* : prefix for naming the result *transfac* file containing all the motifs

```{r, engine = 'bash', eval = FALSE}

## For ChIP-seq derived motifs
Rscript Zinc_Fingers_C2H2_Dogan_2020/pre-process/Rename_C2H2_motifs.R \
  --motif_folder Zinc_Fingers_C2H2_Dogan_2020/data/Zinc_fingers_C2H2_ChIP-seq \
  --logo_folder Zinc_Fingers_C2H2_Dogan_2020/data/Zinc_fingers_C2H2_ChIP-seq/Logos \
  --motif_source ChIP-seq
  
  
## For ChIP-exo derived motifs
Rscript Zinc_Fingers_C2H2_Dogan_2020/pre-process/Rename_C2H2_motifs.R \
  --motif_folder Zinc_Fingers_C2H2_Dogan_2020/data/Zinc_fingers_C2H2_ChIP-exo \
  --logo_folder Zinc_Fingers_C2H2_Dogan_2020/data/Zinc_fingers_C2H2_ChIP-exo/Logos \
  --motif_source ChIP-exo

```

Required software:

  - RSAT
  - R packages:
    - data.table
    - dplyr
    - optparse
    - UniprotR


___


## Pre-processing of *Ciona robusta* motifs derived from SELEX

The motifs were downloaded from the [**Aniseed**](https://www.aniseed.cnrs.fr/aniseed/)
database, in the [**download section**](https://www.aniseed.cnrs.fr/aniseed/download/download_data).

Download the following files:
  
  - *Selex seq section (SELEX-seq best round 6-mer enrichments 4.5 MB)*
  - *SELEX-seq table S1-S3, Nitta et al. 2019 2.7 MB*
  
The first file contains the best PFM of each TF.
The second file contains the annotation of the TFs (IDs, name, family, interprot ID),
we extracted the first sheet and removed the first 2 lines.


Command: 

- *input.motif.folder*   : directory containing the *Aniseed* motifs
- *output.motif.folder*  : directory where the logos and pfms will be stored
- *annotation.table*     : *Aniseed* annotation table
- *organism*             : Ciona_robusta (suffix to name output files)


```{r, engine = 'bash', eval = FALSE}

## For Aniseed (Ciona robusta) SELEX motifs
Rscript Aniseed/pre-process_scripts/Rename_Aniseed_motifs.R \
  --input.motif.folder Aniseed/data/Best_round/pfm_best_round \
  --output.motif.folder Aniseed/data/Best_round \
  --annotation.table Aniseed/data/Best_round/SELEX_project_TableS1_S3_Nitta_et_al_2019.csv \
  --organism ChIP-Ciona_robusta

```

Required software:

  - RSAT
  - R packages:
    - data.table
    - dplyr
    - optparse

___



## Pre-processing of CAP and NCAP SELEX motifs

These motifs were directly taken from the *RSAT* website, in the [**motif databases**](http://rsat-tagc.univ-mrs.fr/rsat/motif_databases/NCAP-SELEX/) section.


___


## Pre-processing of human TF dimer SELEX motifs

These motifs were directly taken from the *RSAT* website, in the [**motif databases**](http://rsat-tagc.univ-mrs.fr/rsat/motif_databases/Jolma_2015/) section.


\pagebreak
# References




